<a href="https://colab.research.google.com/github/Obscuriosity/Programming_For_Data_Analysis/blob/main/ST20214513_CMP7005_PRAC_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Programming for Data Analysis CMP7005**

Tony WIllett ST20214513

In [1]:
!pip install streamlit
!pip install --upgrade streamlit

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 9.1/9.1 MB 30.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.9/6.9 MB 44.5 MB/s eta 0:00:00


In [15]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import streamlit as st
import os
from sklearn.experimental import enable_iterative_imputer
from sklearn.impute import IterativeImputer
import numpy as np

!git config --global user.name "Obscuriosity"
!git config --global user.email "tony.obscuriosity@gmail.com"
username = "Obscuriosity"
repo = "Programming_for_Data_Analysis"
!git clone https://github.com/{username}/{repo}


fatal: destination path 'Programming_for_Data_Analysis' already exists and is not an empty directory.


# Task 1: Data Selection & Handling

You are required to construct a combined dataset for analysis by selecting:

• Select 2 inner (urban) and 2 outer (suburban) stations. provide short justification for your
selection)

▪ Download the four selected datasets.

▪ Import them into your development environment.

▪ Merge them into a single unified dataset for analysis (ensure proper handling of timestamps and station
identifiers).

Choose Locations and upload data.
Monitoring Station:

Dongsi DS - This urban monitoring station seems to be closest to the centre of Beijing

Gucheng GC - This Station is furthest from DS whilst still being an urban station.

Changpingzhen CP - This was chosen because it is an suburban station but is still close to the urban area. It may give information about the influence of the urban area on the local environs.

Huairouzhen HR - This station appears to be the furthest suburban station from the urban centre.

Together these four will give a sort of cross section going form the urban centre, out to the furthest suburbs.


In [5]:
# Load monitoring station data CSV and make pandas dataframes
df_dongsi = pd.read_csv('/content/Programming_for_Data_Analysis/Data/PRSA_Data_Dongsi_20130301-20170228.csv')
df_Gucheng = pd.read_csv('/content/Programming_for_Data_Analysis/Data/PRSA_Data_Changping_20130301-20170228.csv')
df_Changpingzhen = pd.read_csv('/content/Programming_for_Data_Analysis/Data/PRSA_Data_Changping_20130301-20170228.csv')
df_Huairou = pd.read_csv('/content/Programming_for_Data_Analysis/Data/PRSA_Data_Huairou_20130301-20170228.csv')

# Combine dataframes into one
df = pd.concat([df_dongsi, df_Gucheng, df_Changpingzhen, df_Huairou])
df.head()

,No,year,month,day,hour,PM2.5,PM10,SO2,NO2,CO,O3,TEMP,PRES,DEWP,RAIN,wd,WSPM,station
0,1,2013,3,1,0,9.0,9.0,3.0,17.0,300.0,89.0,-0.5,1024.5,-21.4,0.0,NNW,5.7,Dongsi
1,2,2013,3,1,1,4.0,4.0,3.0,16.0,300.0,88.0,-0.7,1025.1,-22.1,0.0,NW,3.9,Dongsi
2,3,2013,3,1,2,7.0,7.0,NaN,17.0,300.0,60.0,-1.2,1025.3,-24.6,0.0,NNW,5.3,Dongsi
3,4,2013,3,1,3,3.0,3.0,5.0,18.0,NaN,NaN,-1.4,1026.2,-25.5,0.0,N,4.9,Dongsi
4,5,2013,3,1,4,3.0,3.0,7.0,NaN,200.0,84.0,-1.9,1027.1,-24.5,0.0,NNW,3.2,Dongsi


# Task 2: Exploratory Data Analysis (EDA)

2.1.Data Understanding
Provide an overview that may include the following, but not limited to:
▪ Number of rows and columns
▪ Column descriptions
▪ Data types
▪ Missing values
▪ Statistical Summary
▪ Initial observations & interpretation

2.2.Data preprocessing:
Perform the necessary data preprocessing steps, including but not limited to handling missing values,
removing duplicate entries, feature engineering (e.g., datetime components, AQI levels), and overall data
cleaning on the main dataset.

2.3.Statistical/Computational Analysis & Visualisation
Perform the necessary steps such as univariate (distribution of pollutants & meteorological variables),
bivariate(e.g. relationships such as PM2.5 vs. Temp, NO2 vs. O3 but not limited to these), and
multivariate analysis (correlation, heatmaps, pairplots), statistical summary, and visualizing the data
(Various charts and graphs, such as bar charts, line charts and scatter plots) that will help in
understanding relationships between variables and to gain important insights from data. Interpret the
key results to demonstrate understanding generated from statistical and visual analysis.

• Explore the dataset however you find meaningful. You may examine different variables, compare
stations, investigate temporal behaviours, or analyse interactions between pollutants and
meteorological factors. Choose the approaches that you believe best help you understand and interpret
the dataset, and present the insights you consider most relevant

In [7]:
# Using Pandas to gain insight into the raw data
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 140256 entries, 0 to 35063
Data columns (total 18 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   No       140256 non-null  int64  
 1   year     140256 non-null  int64  
 2   month    140256 non-null  int64  
 3   day      140256 non-null  int64  
 4   hour     140256 non-null  int64  
 5   PM2.5    137005 non-null  float64
 6   PM10     137762 non-null  float64
 7   SO2      137357 non-null  float64
 8   NO2      135682 non-null  float64
 9   CO       132595 non-null  float64
 10  O3       137233 non-null  float64
 11  TEMP     140079 non-null  float64
 12  PRES     140083 non-null  float64
 13  DEWP     140077 non-null  float64
 14  RAIN     140079 non-null  float64
 15  wd       139596 non-null  object 
 16  WSPM     140107 non-null  float64
 17  station  140256 non-null  object 
dtypes: float64(11), int64(5), object(2)
memory usage: 20.3+ MB


The info function shows that there are 140,256 entries and straight away reveals that not all columns contain the full count of data. We can safely assume that the C0 column, which has the fewest non-null entries (132,595) therefore contains the most null data. As would be expected, the date, time and station name data all contin the full count wheras the actual readings all fall short by varying amounts, this may be due to faulty sensors or any number of reasons.

In [10]:
# some pandas functions to elicit individual data insights
print(df.shape)
print(df.columns)
print(df.dtypes)

(140256, 18)
Index(['No', 'year', 'month', 'day', 'hour', 'PM2.5', 'PM10', 'SO2', 'NO2',
       'CO', 'O3', 'TEMP', 'PRES', 'DEWP', 'RAIN', 'wd', 'WSPM', 'station'],
      dtype='object')
No           int64
year         int64
month        int64
day          int64
hour         int64
PM2.5      float64
PM10       float64
SO2        float64
NO2        float64
CO         float64
O3         float64
TEMP       float64
PRES       float64
DEWP       float64
RAIN       float64
wd          object
WSPM       float64
station     object
dtype: object


The shape command confirms that there are 140,256 entries and 18 columns. The columns and dtypes functions repeat the information gained form the info() function, showing the column names and the data types.

In [11]:
# a look at the null values in more detail
print(df.isnull().sum())

No            0
year          0
month         0
day           0
hour          0
PM2.5      3251
PM10       2494
SO2        2899
NO2        4574
CO         7661
O3         3023
TEMP        177
PRES        173
DEWP        179
RAIN        177
wd          660
WSPM        149
station       0
dtype: int64


The sum of null values command shows that CO indeed contains the most null entries. 7,661 represents just 5.5% of the total 140256 readings. the next highest content of null values is 4,574 and equates to 3.3%. This means that the data is quite complete and that imputing values, even if slightly inaccurate will not adversely affect the conclusions drawn. Nonetheless imputation will be done to preserve dat aintegrity as much as possible.

In [ ]:
# applying multiple imputation techniques to the missing data



In [16]:
# Group columns with missing data into one dataframe
numerical_cols_with_missing = df.select_dtypes(include=np.number).columns[df.select_dtypes(include=np.number).isnull().any()].tolist()

print(f"Numerical columns with missing values: {numerical_cols_with_missing}")

Numerical columns with missing values: []


In [14]:
# Initialize IterativeImputer
imputer = IterativeImputer(max_iter=10, random_state=0)

# Apply imputation to the numerical columns with missing values
df[numerical_cols_with_missing] = imputer.fit_transform(df[numerical_cols_with_missing])

# For wd, impute the most frequent value (mode), the prevailing wind direction
if 'wd' in df.columns and df['wd'].isnull().any():
    mode_wd = df['wd'].mode()[0]
    df['wd'] = df['wd'].fillna(mode_wd)

# CHeck null values
print("Missing values after imputation:")
print(df.isnull().sum())

Missing values after imputation:
No         0
year       0
month      0
day        0
hour       0
PM2.5      0
PM10       0
SO2        0
NO2        0
CO         0
O3         0
TEMP       0
PRES       0
DEWP       0
RAIN       0
wd         0
WSPM       0
station    0
dtype: int64


After applying the `IterativeImputer` for numerical columns and mode imputation for the categorical 'wd' column, all missing values have been handled. The `df.isnull().sum()` output confirms that there are no remaining missing values in the DataFrame.

# Task 3: Model Building:

• After completing all the tasks listed under Task 1 and Task 2, identify and implement the best practices
to build a suitable machine-learning model (e.g., feature scaling, encoding techniques, variable selection,
and parameter optimization).

• Justify your modelling decisions and evaluate model performance using appropriate metrics.


# Task 4: Application Development

Develop an interactive application with a graphical user interface (GUI). The application should include multiple
sections/pages that allow users to explore

• The dataset section,

• Visualization section, and

• Model outputs section.

You may design the structure in any way you find appropriate, but it should enable clear navigation between the

key components of your workflow

# Task 5: Version Control:

• Use GitHub for version control.

• Commit changes regularly with clear, descriptive messages, for example, added PM2.5 prediction
model”, “Created correlation heatmap,” etc.

• Maintain an organised repository structure and include screenshots of:

▫ GitHub commit history

▫ GitHub project repository layout

Links to Papers-
[Yao et al](https://www.mdpi.com/1660-4601/12/10/12264) [Xu&zhang](https://www.sciencedirect.com/science/article/pii/S0301479720301985?via%3Dihub#bib1)